File: ee/ee.go

package main

import (
  "github.com/codegangsta/cli"
)
func main() {
app := cli.NewApp()
app.Name = "EasyEngine Does amazing things"
app.Commands = []cli.Command{
   site_commands,
   stack_commands,
}

app.RunAndExitOnError()
}


var site_commands = cli.Command{
        Name:        "site",
        Usage:       "Site create/delete/update",
        Description: "Site create/delete/update",
        Subcommands: []cli.Command{
            {
                Name:        "create",
                Usage:       "provide website site name",
                Description: "creates site",
                Flags: []cli.Flag{
                    cli.BoolFlag{
                        Name:  "html",
                        Usage: "html site",
                    },
                    cli.BoolFlag{
                        Name:  "php",
                        Usage: "php site",
                    },
                    cli.BoolFlag{
                        Name:  "mysql",
                        Usage: "mysql site",
                    },
                    cli.BoolFlag{
                        Name:  "wp",
                        Usage: "wordpress site",
                    },
                    cli.BoolFlag{
                        Name:  "wpsubdir",
                        Usage: "wordpress multisite subdirectory",
                    },
                    cli.BoolFlag{
                        Name:  "wpsubdomain",
                        Usage: "wordpress multisite subdomain",
                    },
                },
                Action: func(c *cli.Context) {
                  args := c.Args()
                  if args.Present() {
                       if c.Bool("html") {
                        println("site", c.Args().First(), "created with html option" )
                        }
                        if c.Bool("php") {
                            println("site", c.Args().First(), "created with php option")
                        }
                        if c.Bool("mysql") {
                            println("site", c.Args().First(), "created with mysql option")
                        }
                        if c.Bool("wp") {
                            println("site", c.Args().First(), "created with wp option")
                        }
                        if c.Bool("wpsubdir") {
                            println("site", c.Args().First(), "created with wpsubdir option")
                        }
                        if c.Bool("wpsubdomain") {
                            println("site", c.Args().First(), "created with wpsubdomain option")
                        }       
                    } else {
                        println("Please provide site name")
                    }
                },
            },
        },//Site
    }

 var stack_commands = cli.Command{
        Name:      "stack",
        Usage:     "stack install/purge/remove web/mail/admin/mailscanner/all",
        
        Subcommands: []cli.Command{
            {
                Name:        "install",
                Usage:       "Install web/mail/admin/mailscanner/all packages",
                Description: "Installs packages",
                Flags: []cli.Flag{
                    cli.BoolFlag{
                        Name:  "web",
                        Usage: "Install web packages",
                    },
                    cli.BoolFlag{
                        Name:  "mail",
                        Usage: "Install mail packages",
                    },
                    cli.BoolFlag{
                        Name:  "admin",
                        Usage: "Install admin packages",
                    },
                   cli.BoolFlag{
                        Name:  "mailscanner",
                        Usage: "Install mailscanner packages",
                    },
                    cli.BoolFlag{
                        Name:  "all",
                        Usage: "Install all packages",
                    },
                },
                Action: func(c *cli.Context) {
                    println(c.Bool("web"))
                    if c.Bool("web") {
                        println("Installed stack : web")
                    }else if c.Bool("mail") {
                        println("Installed stack : mail")
                    }else if c.Bool("admin") {
                        println("Installed stack : admin")
                    }else if c.Bool("mailscanner") {
                        println("Installed stack : mailscanner")
                    }else if c.Bool("all") {
                        println("Installed stack : all")
                    }
                },
            },{
                Name:        "purge",
                Usage:       "purge web/mail/admin/mailscanner packages",
                Description: "Purge packages",
                Flags: []cli.Flag{
                    cli.BoolFlag{
                        Name:  "web",
                        Usage: "Purge web packages",
                    },
                    cli.BoolFlag{
                        Name:  "mail",
                        Usage: "Purge mail packages",
                    },
                    cli.BoolFlag{
                        Name:  "admin",
                        Usage: "Purge admin packages",
                    },
                   cli.BoolFlag{
                        Name:  "mailscanner",
                        Usage: "Purge mailscanner packages",
                    },
                    cli.BoolFlag{
                        Name:  "all",
                        Usage: "Purge all packages",
                    },
                    
                },
                Action: func(c *cli.Context) {
                    if c.Bool("web") {
                        println("Purged stack : web")
                    }else if c.Bool("mail") {
                        println("Purged stack : mail")
                    }else if c.Bool("admin") {
                        println("Purged stack : admin")
                    }else if c.Bool("mailscanner") {
                        println("Purged stack : mailscanner")
                    }else if c.Bool("all") {
                        println("Purged stack : all")
                    }
                },//end of Action
            },{
                Name:        "remove",
                Usage:       "remove web/mail/admin/mailscanner/all packages",
                Description: "remove packages",
                Flags: []cli.Flag{
                    cli.BoolFlag{
                        Name:  "web",
                        Usage: "Remove web packages",
                    },
                    cli.BoolFlag{
                        Name:  "mail",
                        Usage: "Remove mail packages",
                    },
                    cli.BoolFlag{
                        Name:  "admin",
                        Usage: "Remove admin packages",
                    },
                   cli.BoolFlag{
                        Name:  "mailscanner",
                        Usage: "Remove mailscanner packages",
                    },
                   cli.BoolFlag{
                        Name:  "all",
                        Usage: "Remove all packages",
                    },
                },
                Action: func(c *cli.Context) {
                    if c.Bool("web") {
                        println("Removed stack : web")
                    }else if c.Bool("mail") {
                        println("Removed stack : mail")
                    }else if c.Bool("admin") {
                        println("Removed stack : admin")
                    }else if c.Bool("mailscanner") {
                        println("Removed stack : mailscanner")
                    }else if c.Bool("all") {
                        println("Removed stack : all")
                    }
                },//end of Action
            },
        },
    }

File: python/python-cement/ee.py
#!/usr/bin/env python3.4

from cement.core import foundation, controller, handler

# define application controllers
class EEBaseController(controller.CementBaseController):
		class Meta:
				label = 'base'
				interface = controller.IController
				description = "EasyEngine Does Amazing Things"

		@controller.expose(help="EayEngine command", hide=True)
		def default(self):
				self.app.args.parse_args(['--help'])

class Site(controller.CementBaseController):
		"""This controller commands are 'stacked' onto the base controller."""

		class Meta:
				label = 'site'
				interface = controller.IController
				stacked_on = 'base'
				stacked_type = 'nested'
				description = "Site commands are work under this controller"

## argument selection is an issue
class Create(controller.CementBaseController):
		"""This controller commands are 'stacked' onto the site controller."""

		class Meta:
				label = 'create'
				interface = controller.IController
				stacked_on = 'site'
				stacked_type = 'nested'
				description = "create commands are listed under this controller"
				arguments = [
						(['site_name'], dict(help="site domain name like example.com")),
						(['--html'], dict(help="create html site example.com", action='store_true')),
						(['--php'], dict(help="create php site example.com", action='store_true')),
						(['--mysql'], dict(help="create mysql site example.com", action='store_true')),
						(['--wp'], dict(help="create wp site example.com", action='store_true')),
						(['--wpsubdir'], dict(help="create wpsubdir site example.com", action='store_true')),
						(['--wpsubdomain'], dict(help="create wpsubdomain site example.com", action='store_true')),
						]

		@controller.expose(help="command under the base namespace")
		def default(self):
				if self.app.pargs.html:
						print ("site", self.app.pargs.site_name ,"created with html option")
				if self.app.pargs.php:
						print ("site", self.app.pargs.site_name, "created with php option")
				if self.app.pargs.mysql:
						print ("site", self.app.pargs.site_name, "created with mysql option")
				if self.app.pargs.wp:
						print ("site", self.app.pargs.site_name, "created with wp option")
				if self.app.pargs.wpsubdir:
						print ("site", self.app.pargs.site_name, "created with wpsubdir option")
				if self.app.pargs.wpsubdomain:
						print ("site", self.app.pargs.site_name, "created with wpsubdomain option")


class Stack(controller.CementBaseController):
		"""This controller commands are *not* 'stacked' onto the base controller."""

		class Meta:
				label = 'stack'
				interface = controller.IController
				stacked_on = 'base'
				stacked_type = 'nested'
				description = "Stack commands are listed under this controller"



class Install(controller.CementBaseController):
		"""This controller commands are *not* 'stacked' onto the base controller."""

		class Meta:
				label = 'install'
				interface = controller.IController
				stacked_on = 'stack'
				stacked_type = 'nested'
				description = "install commands	are listed under this controller"
				arguments = [
						(['--web'], dict(help="Install web stack", action='store_true')),
						(['--admin'], dict(help="Install admin stack", action='store_true')),
						(['--mail'], dict(help="Install mail stack", action='store_true')),
						(['--mailscanner'], dict(help="Install mailscanner stack", action='store_true')),
						(['--all'], dict(help="Install all stack", action='store_true')),
						]

		@controller.expose(help="default command for third_controller", hide=True)
		def default(self):
				
				if self.app.pargs.web:
								print("Installed stack : web")
				if self.app.pargs.admin:
								print( "Installed stack : admin")
				if self.app.pargs.mail:
								print( "Installed stack : mail")
				if self.app.pargs.mailscanner:
								print( "Installed stack : mailscanner")
				if self.app.pargs.all:
								print( "Installed stack : all")

try:
		# create the application
		app = foundation.CementApp('ee')

		# register non-base controllers
		handler.register(EEBaseController)
		handler.register(Site)
		handler.register(Stack)
		handler.register(Create)
		handler.register(Install)


		# setup the application
		app.setup()

		app.run()
finally:
		app.close()
